### ISSUE TITLE
Generalize WebID-OIDC protocol beyond WebID

### ISSUE THREAD

**Issue by acoburn** [solid/solid-oidc#35]
WebID provides a structured way to identify and refer to agents, but there are now other structured forms of identification, such as [Decentralized Identifiers (DIDs)](https://w3c-ccg.github.io/did-spec/) and [Verifiable Credentials](https://www.w3.org/TR/vc-data-model/). The name and protocol of WebID-OIDC is very WebID-centric, and perhaps the protocol should be made flexible enough to accommodate other identity structures, such as those mentioned here but also those that haven't been invented yet.

**Comment by acoburn** [comment]
This issue has been open for nearly two years. It seems that the general consensus is that we would like to represent agent identity as some form of an identifier that can be dereferenced as an RDF document. That general opinion would include, but not be limited to, DID and WebID.

As the specification currently stands, there is a tight connection to WebID. Unfortunately, the WebID draft specification is still a draft with no clear movement toward recommendation status.

I would like to make a specific proposal that the Solid-OIDC specification not normatively require WebID as it does today. It would not exclude the use of WebID, but by being more generic, it avoids some of the problematic portions of the current WebID draft. In particular:

* the current use of "Client WebID" in Solid-OIDC conflicts with the WebID draft's requirement to serialize resources as Turtle
* the current requirement of WebID URLs to use the HTTP(S) scheme excludes DIDs

I opened https://github.com/w3c/WebID/issues/3 in an attempt to move the WebID document in a more flexible direction, but the initial comments do not appear to be going in that direction.

**Comment by bblfish** [comment]
Yes, it would be good to have others in the community give feedback on how they think we could proceed with the issue brought up by @acoburn . 

Note that the more you generalise, the more complicated things become in terms of engineering and debugging. Oddly that does not seem to be the case in Mathematics where short proofs starting from very abstract principles are valued, because they are for those in the know, easier to oversee. See the [recent talk by Kevin Buzzard](https://twitter.com/bblfish/status/1426188329163403268) on formalizing mathematics in Lean.  Even there though, it helps often to rewrite proofs from simpler premises to help explain the concepts to undergraduate mathematics students.

In the web space, on these issues,  leaving formats and protocols open, creates requirement on clients to implement parsers for those formats and protocols. This is because we are aiming at interoperability. 
So say you want DID's: then you potentially require every implementor to implement all DID protocols. 
So it looks all nice and open to everyone, but you can massively increase the complexity of adoption of Solid in the process of doing so. There are then questions as to how much resources the community has at this point. As time goes by, more resources may be available, and things that are complicated now, may be affordable then.

Indeed, your argument @acoburn, is built on the use case that people have trouble deploying WebID because they are using github to publish their code which makes content-negotiation difficult or impossible. In order to overcome that problem, you are arguing that developers should prepare to deal with all the generality that DIDs and multiple mime types require.

There must be more pragmatic ways to deal with this issue. I did [propose one](https://github.com/w3c/WebID/issues/3#issuecomment-898586260). I have often found that sleeping over an issue and listening to others brings up new options.

**Comment by acoburn** [comment]
> your argument @acoburn, is built on the use case that people have trouble deploying WebID because they are using github to publish their code which makes content-negotiation difficult or impossible

That is one problem. There are others.

Another problem is that the WebID draft is just a draft. Depending on a draft is problematic because that dependency can change at any time.

What I am suggesting is that we think about WebID in the same way Solid thinks about LDP: it is possible to use it but it is not a hard dependency. By removing the normative dependency on the WebID draft, it means that we are able to build Solid-OIDC on a much more stable foundation, especially one that allows for future growth in the direction of DIDs (that direction is not possible with the strict dependency on WebID).

In fact, the Solid-OIDC document is attempting to be very specific about the sort of thing that matters to OIDC: serialization format, but that conflicts with the WebID specification. As things currently stand with Solid-OIDC w.r.t WebID we are left with having to hand-wave about serialization requirements, effectively saying: "just ignore that part of the WebID spec". That's not a good position to be in. The only real solutions there are to: (a) change the WebID spec (b) change the Solid-OIDC spec or (c) eliminate the hard dependency on WebID

**Comment by bblfish** [comment]
>Another problem is that the WebID draft is just a draft.

Well so is Solid-OIDC and most of the specs that are interesting. Usually when something is no longer a draft, it is because it has very wide adoption. 

> (a) change the WebID spec (b) change the Solid-OIDC spec or (c) eliminate the hard dependency on WebID

You opened an issue on the WebID repository just a couple of hours ago. That repository was opened to allow the spec to evolve to new requirements. Perhaps if we wait for others to make suggestions on how to proceed you will be able to make a more informed decision.

**Comment by acoburn** [comment]
>> Another problem is that the WebID draft is just a draft.

> Well so is Solid-OIDC

Indeed. Solid-OIDC is also actively moving toward a more stable form. We do not want it to always be a draft. Part of moving in that direction means resolving these sorts of long-standing issues.

> You opened an issue on the WebID repository just a couple of hours ago.

True. But from the initial responses, I did not get the sense that the WebID specification will be changing any time soon. The specific concerns I raised were effectively discarded. That did not inspire much confidence in there being any changes. From what I can tell, the WebID draft has not changed since March 5, 2014.

Please remember: I am not suggesting that we _exclude_ the use of WebID, rather that we not depend on it directly, as we currently do.

**Comment by bblfish** [comment]
Note that you can not generalise WebID to DIDs, because `did` is a URI schema that does not encompass `http` or `https`. You will need to have the disjoint union of WebID and DIDs if you want to be open to DIDs. Unless you want to move the OIDC protocol totally away from HTTP.

**Comment by acoburn** [comment]
I am not suggesting a generalization of WebID to include DIDs. You are correct that such a thing is not possible.

I am suggesting that Solid-OIDC identify agents with URIs that can be dereferenced as RDF documents. Examples of such URIs include both WebID and DID.

**Comment by acoburn** [comment]
For example, in the [Solid Web Access Control specification draft](https://solid.github.io/web-access-control-spec/#terminology), it makes reference to WebID without depending on it:

> agent
> An agent is a person, social entity or software identified by a URI, e.g., a WebID denotes an agent [WEBID].

Solid-OIDC could do something similar, though I would also include DID as an example:

> Agent
> An agent is a person, social entity or software identified by a URI, e.g., a DID or WebID can denote an agent.

Contrast that with the current text:

> WebID as defined by [WebID]
> A WebID is a URI with an HTTP or HTTPS scheme which denotes an Agent (Person, Organization, Group, Device, etc.).

**Comment by bblfish** [comment]
> a DID or WebID

yes, that is what I mean by [disjoint union](https://en.wikipedia.org/wiki/Disjoint_union).

**Comment by acoburn** [comment]
> yes, that is what I mean by disjoint union.

Exactly. And so long as Solid-OIDC requires WebID, it cannot support DID

**Comment by bblfish** [comment]
Ok, I am all for you supporting DIDs.
But supporting DID does not solve your mime type problem with WebID.

**Comment by acoburn** [comment]
I am not suggesting that we _replace_ WebID with DIDs. Indeed, that would not solve the problem.

Rather, I am suggesting that Solid-OIDC stay independent of all that. You can use a WebID or a DID or whatever comes next.

That _would_ solve the mime type problem.
